File: mobile-app/lib/data/bike_computer_repository.dart
import 'dart:async';
import 'dart:typed_data';

import '../core/app_error.dart';
import '../core/result.dart';
import '../domain/entities/companion_models.dart';
import '../domain/entities/models.dart';
import '../domain/validators/config_validator.dart';
import 'ble/ble_transport.dart';
import 'protocol/ble_uuids.dart';
import 'protocol/protocol_codecs.dart';

abstract interface class BikeComputerRepository {
  Stream<BleLinkState> get connectionState;
  Stream<Telemetry> get telemetry;
  Stream<DeviceConfig> get configUpdates;
  Stream<int> get rssi;

  Future<Result<DeviceInfo>> readDeviceInfo();
  Future<Result<DeviceConfig>> readConfig();
  Future<Result<Telemetry>> readTelemetry();
  Future<Result<DiagnosticSnapshot>> getDiagnostic();
  Future<Result<ErrorLogBatch>> readErrorLog();
  Future<Result<void>> writeConfig(DeviceConfig config);
  Future<Result<void>> writeCompanionSnapshot(CompanionSnapshot snapshot);
  Future<Result<CommandResult>> sendCommand(DeviceCommand command);
  Future<Result<void>> setOdometerMeters(int odometerM);
  Future<void> setTelemetrySubscribed(bool value);
  Future<void> disconnect();
}

class BikeComputerRepositoryImpl implements BikeComputerRepository {
  BikeComputerRepositoryImpl(this._transport);

  final BleTransport _transport;
  final _connectionState = StreamController<BleLinkState>.broadcast();
  final _telemetry = StreamController<Telemetry>.broadcast();
  final _configUpdates = StreamController<DeviceConfig>.broadcast();
  final _rssi = StreamController<int>.broadcast();
  final _commandResults = StreamController<CommandResult>.broadcast();
  final List<StreamSubscription<List<int>>> _subscriptions = [];
  StreamSubscription<List<int>>? _telemetrySubscription;
  Timer? _rssiTimer;
  DeviceConfig? _latestConfig;
  Future<void> _operationTail = Future<void>.value();

  @override
  Stream<BleLinkState> get connectionState => _connectionState.stream;
  @override
  Stream<Telemetry> get telemetry => _telemetry.stream;
  @override
  Stream<DeviceConfig> get configUpdates => _configUpdates.stream;
  @override
  Stream<int> get rssi => _rssi.stream;

  void noteLinkState(BleLinkState value) => _connectionState.add(value);

  Future<void> start() async {
    final configSubscription = _transport
        .subscribe(BleUuids.configRead)
        .listen(_handleConfigBytes, onError: _configUpdates.addError);
    final resultSubscription = _transport
        .subscribe(BleUuids.commandResult)
        .listen(_handleResultBytes, onError: _commandResults.addError);
    _subscriptions.addAll([configSubscription, resultSubscription]);
    await setTelemetrySubscribed(true);
    _rssiTimer = Timer.periodic(const Duration(seconds: 1), (_) async {
      try {
        _rssi.add(await _transport.readRssi());
      } on Object {
        // RSSI is auxiliary; connection events remain the source of truth.
      }
    });
  }

  void _handleConfigBytes(List<int> bytes) {
    final config = ProtocolCodecs.decodeConfig(bytes);
    _latestConfig = config;
    _configUpdates.add(config);
  }

  void _handleResultBytes(List<int> bytes) {
    _commandResults.add(ProtocolCodecs.decodeCommandResult(bytes));
  }

  void _handleTelemetryBytes(List<int> bytes) {
    _telemetry.add(ProtocolCodecs.decodeTelemetry(bytes));
  }

  Future<Result<T>> _guard<T>(Future<T> Function() body) async {
    try {
      return Success(await body());
    } on AppError catch (error) {
      return Failure(error);
    } on ProtocolCodecException catch (error) {
      return Failure(AppErrors.malformed(error.message, error));
    } on Object catch (error) {
      return Failure(AppErrors.unknown(error));
    }
  }

  Future<Result<T>> _enqueue<T>(Future<Result<T>> Function() body) {
    final completer = Completer<Result<T>>();
    _operationTail = _operationTail.then((_) async {
      try {
        completer.complete(await body());
      } on Object catch (error) {
        completer.complete(Failure<T>(AppErrors.unknown(error)));
      }
    });
    return completer.future;
  }

  @override
  Future<Result<DeviceInfo>> readDeviceInfo() => _guard(
    () async => ProtocolCodecs.decodeDeviceInfo(
      await _transport.read(BleUuids.deviceInfo),
    ),
  );

  @override
  Future<Result<DeviceConfig>> readConfig() => _guard(() async {
    final value = ProtocolCodecs.decodeConfig(
      await _transport.read(BleUuids.configRead),
    );
    _latestConfig = value;
    _configUpdates.add(value);
    return value;
  });

  @override
  Future<Result<Telemetry>> readTelemetry() => _guard(() async {
    final value = ProtocolCodecs.decodeTelemetry(
      await _transport.read(BleUuids.telemetry),
    );
    _telemetry.add(value);
    return value;
  });

  @override
  Future<Result<DiagnosticSnapshot>> getDiagnostic() => _enqueue(() async {
    final response = _nextResult(DeviceCommandId.getDiagnostic);
    try {
      await _transport.writeWithResponse(
        BleUuids.command,
        ProtocolCodecs.encodeCommand(
          const DeviceCommand(id: DeviceCommandId.getDiagnostic),
        ),
      );
      final result = await response;
      if (result.status != CommandStatus.ok) {
        return Failure<DiagnosticSnapshot>(_resultError(result));
      }
      if (result.payload.length != ProtocolCodecs.diagnosticPayloadSize) {
        return Failure<DiagnosticSnapshot>(
          AppErrors.malformed(
            'Diagnostic payload: ожидалось '
            '${ProtocolCodecs.diagnosticPayloadSize} байт, '
            'получено ${result.payload.length}',
          ),
        );
      }
      return Success(ProtocolCodecs.decodeDiagnostic(result.payload));
    } on TimeoutException {
      return const Failure<DiagnosticSnapshot>(
        AppFailure(
          kind: AppErrorKind.writeTimeout,
          message: 'Устройство не ответило на GET_DIAGNOSTIC',
          action: 'Повторить',
        ),
      );
    } on Object catch (error) {
      response.ignore();
      return Failure<DiagnosticSnapshot>(AppErrors.unknown(error));
    }
  });

  @override
  Future<Result<ErrorLogBatch>> readErrorLog() => _guard(() async {
    try {
      return ProtocolCodecs.decodeErrorLog(
        await _transport.read(BleUuids.errorLog),
      );
    } on StateError {
      return const ErrorLogBatch(entries: <ErrorLogEntry>[]);
    }
  });

  Future<CommandResult> _nextResult(DeviceCommandId command) => _commandResults
      .stream
      .firstWhere((result) => result.commandId == command)
      .timeout(const Duration(seconds: 3));

  AppError _resultError(CommandResult result) => switch (result.status) {
    CommandStatus.range => AppFailure(
      kind: AppErrorKind.deviceRejected,
      message:
          'Устройство отклонило поле конфигурации (offset ${result.detail})',
      fieldId: result.detail,
    ),
    CommandStatus.notPaired => AppErrors.notPaired,
    CommandStatus.storage => const AppFailure(
      kind: AppErrorKind.storageError,
      message: 'Устройство не смогло сохранить настройки',
      action: 'Повторить',
    ),
    CommandStatus.tokenExpired => const AppFailure(
      kind: AppErrorKind.tokenExpired,
      message: 'Время подтверждения истекло',
      action: 'Начать заново',
    ),
    CommandStatus.hardware => const AppFailure(
      kind: AppErrorKind.hardware,
      message: 'Устройство сообщило об аппаратной ошибке',
    ),
    CommandStatus.notSupported => const AppFailure(
      kind: AppErrorKind.unsupported,
      message: 'Функция отсутствует в этой версии прошивки',
    ),
    _ => AppFailure(
      kind: AppErrorKind.unknown,
      message: 'Устройство отклонило операцию: ${result.status.name}',
    ),
  };

  @override
  Future<Result<void>> writeConfig(DeviceConfig config) => _enqueue(() async {
    final validation = ConfigValidator.validate(config);
    if (!validation.isValid) {
      final first = validation.issues.first;
      return Failure<void>(
        AppErrors.validation(first.message, fieldId: first.fieldId),
      );
    }

    for (var attempt = 0; attempt <= 3; attempt++) {
      final response = _nextResult(DeviceCommandId.configWrite);
      try {
        await _transport.writeWithResponse(
          BleUuids.configWrite,
          ProtocolCodecs.encodeConfig(config),
        );
        final result = await response;
        if (result.status == CommandStatus.busy && attempt < 3) {
          await Future<void>.delayed(const Duration(seconds: 1));
          continue;
        }
        if (result.status != CommandStatus.ok) {
          return Failure<void>(_resultError(result));
        }
        if (_latestConfig != config) {
          try {
            await _configUpdates.stream
                .firstWhere((value) => value == config)
                .timeout(const Duration(seconds: 3));
          } on TimeoutException {
            final reread = await readConfig();
            if (reread case Success<DeviceConfig>(
              value: final value,
            ) when value == config) {
              return const Success<void>(null);
            }
            return const Failure<void>(AppErrors.writeTimeout);
          }
        }
        return const Success<void>(null);
      } on TimeoutException {
        final reread = await readConfig();
        if (reread case Success<DeviceConfig>(
          value: final value,
        ) when value == config) {
          return const Success<void>(null);
        }
        return const Failure<void>(AppErrors.writeTimeout);
      } on Object catch (error) {
        response.ignore();
        return Failure<void>(AppErrors.unknown(error));
      }
    }
    return const Failure<void>(
      AppFailure(
        kind: AppErrorKind.busy,
        message: 'Устройство занято. Повторите операцию',
        action: 'Повторить',
      ),
    );
  });

  Future<Result<CommandResult>> _sendCommandDirect(
    DeviceCommand command,
  ) async {
    for (var attempt = 0; attempt <= 3; attempt++) {
      final response = _nextResult(command.id);
      try {
        await _transport.writeWithResponse(
          BleUuids.command,
          ProtocolCodecs.encodeCommand(command),
        );
        final result = await response;
        if (result.status == CommandStatus.busy && attempt < 3) {
          await Future<void>.delayed(const Duration(seconds: 1));
          continue;
        }
        if (result.status == CommandStatus.ok ||
            result.status == CommandStatus.needsConfirm) {
          return Success(result);
        }
        return Failure<CommandResult>(_resultError(result));
      } on TimeoutException {
        return const Failure<CommandResult>(
          AppFailure(
            kind: AppErrorKind.writeTimeout,
            message: 'Устройство не ответило на команду',
            action: 'Повторить',
          ),
        );
      } on Object catch (error) {
        response.ignore();
        return Failure<CommandResult>(AppErrors.unknown(error));
      }
    }
    return const Failure<CommandResult>(
      AppFailure(
        kind: AppErrorKind.busy,
        message: 'Устройство занято. Повторите команду',
        action: 'Повторить',
      ),
    );
  }

  @override
  Future<Result<void>> writeCompanionSnapshot(CompanionSnapshot snapshot) =>
      _enqueue(() async {
        await _transport.writeWithResponse(
          BleUuids.companionWrite,
          ProtocolCodecs.encodeCompanion(snapshot),
        );
        return const Success<void>(null);
      });

  @override
  Future<Result<CommandResult>> sendCommand(DeviceCommand command) =>
      _enqueue(() => _sendCommandDirect(command));

  @override
  Future<Result<void>> setOdometerMeters(int odometerM) => _enqueue(() async {
    final requestPayload = ByteData(4)..setUint32(0, odometerM, Endian.little);
    final request = DeviceCommand(
      id: DeviceCommandId.setOdometer,
      payload: requestPayload.buffer.asUint8List().toList(),
    );
    final stepOne = await _sendCommandDirect(request);
    if (stepOne case Failure<CommandResult>(:final error)) {
      return Failure<void>(error);
    }
    final needs = (stepOne as Success<CommandResult>).value;
    if (needs.status != CommandStatus.needsConfirm) {
      return needs.status == CommandStatus.ok
          ? const Success<void>(null)
          : Failure<void>(_resultError(needs));
    }
    final tokenBytes = ByteData(4)..setUint32(0, needs.token, Endian.little);
    final confirmPayload = <int>[
      ...requestPayload.buffer.asUint8List(),
      ...tokenBytes.buffer.asUint8List(),
    ];
    final confirm = DeviceCommand(
      id: DeviceCommandId.setOdometer,
      hasToken: true,
      payload: confirmPayload,
    );
    final stepTwo = await _sendCommandDirect(confirm);
    if (stepTwo case Failure<CommandResult>(:final error)) {
      return Failure<void>(error);
    }
    final result = (stepTwo as Success<CommandResult>).value;
    if (result.status == CommandStatus.ok) {
      return const Success<void>(null);
    }
    return Failure<void>(_resultError(result));
  });

  @override
  Future<void> setTelemetrySubscribed(bool value) async {
    if (!value) {
      await _telemetrySubscription?.cancel();
      _telemetrySubscription = null;
      return;
    }
    if (_telemetrySubscription != null) return;
    _telemetrySubscription = _transport
        .subscribe(BleUuids.telemetry)
        .listen(_handleTelemetryBytes, onError: _telemetry.addError);
  }

  @override
  Future<void> disconnect() => _transport.disconnect();

  Future<void> dispose() async {
    _rssiTimer?.cancel();
    await _telemetrySubscription?.cancel();
    for (final subscription in _subscriptions) {
      await subscription.cancel();
    }
    await _connectionState.close();
    await _telemetry.close();
    await _configUpdates.close();
    await _rssi.close();
    await _commandResults.close();
  }
}

File: web-app/src/services/bikeComputerRepository.ts
import {
  type AppError,
  AppErrors,
  failure,
  type Result,
  success,
} from '../core/result';
import {
  type CommandResult,
  type CompanionSnapshot,
  configEquals,
  type DeviceCommand,
  type DeviceConfig,
  type DeviceInfo,
  type DiagnosticSnapshot,
  type ErrorLogBatch,
  type Telemetry,
  type DeviceCommandId,
} from '../domain/types';
import { validateConfig } from '../domain/configValidator';
import {
  decodeCommandResult,
  decodeConfig,
  decodeDeviceInfo,
  decodeDiagnostic,
  decodeErrorLog,
  decodeTelemetry,
  encodeCommand,
  encodeCompanion,
  encodeConfig,
  ProtocolCodecError,
  SIZES,
} from '../protocol/codecs';
import { BleUuids } from '../protocol/uuids';
import type { BleLinkState, BleTransport } from '../transport/bleTransport';

type Listener<T> = (value: T) => void;

export class BikeComputerRepository {
  private connectionListeners = new Set<Listener<BleLinkState>>();
  private telemetryListeners = new Set<Listener<Telemetry>>();
  private configListeners = new Set<Listener<DeviceConfig>>();
  private rssiListeners = new Set<Listener<number>>();
  private commandResultListeners = new Set<Listener<CommandResult>>();

  private subscriptions: Array<() => void> = [];
  private telemetryAbort: AbortController | null = null;
  private rssiTimer: ReturnType<typeof setInterval> | null = null;
  private latestConfig: DeviceConfig | null = null;
  private operationTail: Promise<void> = Promise.resolve();

  constructor(private readonly transport: BleTransport) {}

  onConnection(listener: Listener<BleLinkState>) {
    this.connectionListeners.add(listener);
    return () => this.connectionListeners.delete(listener);
  }

  onTelemetry(listener: Listener<Telemetry>) {
    this.telemetryListeners.add(listener);
    return () => this.telemetryListeners.delete(listener);
  }

  onConfig(listener: Listener<DeviceConfig>) {
    this.configListeners.add(listener);
    return () => this.configListeners.delete(listener);
  }

  onRssi(listener: Listener<number>) {
    this.rssiListeners.add(listener);
    return () => this.rssiListeners.delete(listener);
  }

  noteLinkState(state: BleLinkState) {
    for (const l of this.connectionListeners) l(state);
  }

  async start(): Promise<void> {
    const configAbort = new AbortController();
    const resultAbort = new AbortController();
    void this.pumpSubscribe(BleUuids.configRead, configAbort.signal, (bytes) => {
      const config = decodeConfig(bytes);
      this.latestConfig = config;
      for (const l of this.configListeners) l(config);
    });
    void this.pumpSubscribe(BleUuids.commandResult, resultAbort.signal, (bytes) => {
      const result = decodeCommandResult(bytes);
      for (const l of this.commandResultListeners) l(result);
    });
    this.subscriptions.push(() => {
      configAbort.abort();
      resultAbort.abort();
    });
    await this.setTelemetrySubscribed(true);
    this.rssiTimer = setInterval(async () => {
      try {
        const rssi = await this.transport.readRssi();
        for (const l of this.rssiListeners) l(rssi);
      } catch {
        // auxiliary
      }
    }, 1000);
  }

  async dispose(): Promise<void> {
    if (this.rssiTimer) clearInterval(this.rssiTimer);
    this.telemetryAbort?.abort();
    for (const unsub of this.subscriptions) unsub();
    this.subscriptions = [];
  }

  private async pumpSubscribe(
    uuid: string,
    signal: AbortSignal,
    handler: (bytes: Uint8Array) => void,
  ) {
    try {
      for await (const bytes of this.transport.subscribe(uuid)) {
        if (signal.aborted) break;
        handler(bytes);
      }
    } catch {
      // subscription ended
    }
  }

  private guard<T>(body: () => Promise<T>): Promise<Result<T>> {
    return body()
      .then((value) => success(value))
      .catch((error: unknown) => {
        if (error instanceof ProtocolCodecError) {
          return failure<T>(AppErrors.malformed(error.message));
        }
        return failure<T>(AppErrors.unknown(error));
      });
  }

  private enqueue<T>(body: () => Promise<Result<T>>): Promise<Result<T>> {
    const run = this.operationTail.then(body);
    this.operationTail = run.then(() => undefined).catch(() => undefined);
    return run;
  }

  readDeviceInfo(): Promise<Result<DeviceInfo>> {
    return this.guard(async () =>
      decodeDeviceInfo(await this.transport.read(BleUuids.deviceInfo)),
    );
  }

  readConfig(): Promise<Result<DeviceConfig>> {
    return this.guard(async () => {
      const value = decodeConfig(await this.transport.read(BleUuids.configRead));
      this.latestConfig = value;
      for (const l of this.configListeners) l(value);
      return value;
    });
  }

  readTelemetry(): Promise<Result<Telemetry>> {
    return this.guard(async () => {
      const value = decodeTelemetry(await this.transport.read(BleUuids.telemetry));
      for (const l of this.telemetryListeners) l(value);
      return value;
    });
  }

  readErrorLog(): Promise<Result<ErrorLogBatch>> {
    return this.guard(async () => {
      try {
        return decodeErrorLog(await this.transport.read(BleUuids.errorLog));
      } catch {
        return { entries: [] };
      }
    });
  }

  getDiagnostic(): Promise<Result<DiagnosticSnapshot>> {
    return this.enqueue(async () => {
      const response = this.nextResult('getDiagnostic');
      try {
        await this.transport.writeWithResponse(
          BleUuids.command,
          encodeCommand({ structVersion: 1, id: 'getDiagnostic', hasToken: false, payload: [] }),
        );
        const result = await response;
        if (result.status !== 'ok') return failure<DiagnosticSnapshot>(this.resultError(result));
        if (result.payload.length !== SIZES.diagnosticPayload) {
          return failure<DiagnosticSnapshot>(
            AppErrors.malformed(
              `Diagnostic payload: ожидалось ${SIZES.diagnosticPayload} байт`,
            ),
          );
        }
        return success(decodeDiagnostic(Uint8Array.from(result.payload)));
      } catch (error) {
        return failure<DiagnosticSnapshot>(AppErrors.unknown(error));
      }
    });
  }

  writeConfig(config: DeviceConfig): Promise<Result<void>> {
    return this.enqueue(async () => {
      const validation = validateConfig(config);
      if (!validation.isValid) {
        const first = validation.issues[0];
        return failure<void>(AppErrors.validation(first.message, first.fieldId));
      }
      for (let attempt = 0; attempt <= 3; attempt++) {
        const response = this.nextResult('configWrite');
        try {
          await this.transport.writeWithResponse(BleUuids.configWrite, encodeConfig(config));
          const result = await response;
          if (result.status === 'busy' && attempt < 3) {
            await delay(1000);
            continue;
          }
          if (result.status !== 'ok') return failure<void>(this.resultError(result));
          if (this.latestConfig && !configEquals(this.latestConfig, config)) {
            await this.waitForConfig(config, 3000);
          }
          return success(undefined);
        } catch (error) {
          const reread = await this.readConfig();
          if (reread.ok && configEquals(reread.value, config)) return success(undefined);
          return failure<void>(AppErrors.unknown(error));
        }
      }
      return failure<void>({
        kind: 'busy',
        message: 'Устройство занято. Повторите операцию',
        action: 'Повторить',
      });
    });
  }

  writeCompanionSnapshot(snapshot: CompanionSnapshot): Promise<Result<void>> {
    return this.enqueue(async () => {
      try {
        await this.transport.writeWithResponse(BleUuids.companionWrite, encodeCompanion(snapshot));
        return success(undefined);
      } catch (error) {
        return failure<void>(AppErrors.unknown(error));
      }
    });
  }

  sendCommand(command: DeviceCommand): Promise<Result<CommandResult>> {
    return this.enqueue(() => this.sendCommandDirect(command));
  }

  setOdometerMeters(odometerM: number): Promise<Result<void>> {
    const payload = new Uint8Array(4);
    new DataView(payload.buffer).setUint32(0, odometerM, true);
    return this.enqueue(() =>
      this.sendDangerousCommand({
        structVersion: 1,
        id: 'setOdometer',
        hasToken: false,
        payload: Array.from(payload),
      }),
    );
  }

  rebootDevice(): Promise<Result<void>> {
    return this.enqueue(() =>
      this.sendDangerousCommand({
        structVersion: 1,
        id: 'reboot',
        hasToken: false,
        payload: [],
      }),
    );
  }

  async setTelemetrySubscribed(value: boolean): Promise<void> {
    if (!value) {
      this.telemetryAbort?.abort();
      this.telemetryAbort = null;
      return;
    }
    if (this.telemetryAbort) return;
    const abort = new AbortController();
    this.telemetryAbort = abort;
    void this.pumpSubscribe(BleUuids.telemetry, abort.signal, (bytes) => {
      const telemetry = decodeTelemetry(bytes);
      for (const l of this.telemetryListeners) l(telemetry);
    });
  }

  disconnect(): Promise<void> {
    return this.transport.disconnect();
  }

  private async sendCommandDirect(command: DeviceCommand): Promise<Result<CommandResult>> {
    for (let attempt = 0; attempt <= 3; attempt++) {
      const response = this.nextResult(command.id);
      try {
        await this.transport.writeWithResponse(BleUuids.command, encodeCommand(command));
        const result = await response;
        if (result.status === 'busy' && attempt < 3) {
          await delay(1000);
          continue;
        }
        if (result.status === 'ok' || result.status === 'needsConfirm') {
          return success(result);
        }
        return failure<CommandResult>(this.resultError(result));
      } catch (error) {
        return failure<CommandResult>(AppErrors.unknown(error));
      }
    }
    return failure<CommandResult>({
      kind: 'busy',
      message: 'Устройство занято. Повторите команду',
      action: 'Повторить',
    });
  }

  private async sendDangerousCommand(request: DeviceCommand): Promise<Result<void>> {
    const stepOne = await this.sendCommandDirect(request);
    if (!stepOne.ok) return failure(stepOne.error);
    const needs = stepOne.value;
    if (needs.status !== 'needsConfirm') {
      return needs.status === 'ok' ? success(undefined) : failure(this.resultError(needs));
    }
    const tokenBytes = new Uint8Array(4);
    new DataView(tokenBytes.buffer).setUint32(0, needs.token, true);
    const confirmPayload = [...request.payload, ...tokenBytes];
    const confirm: DeviceCommand = {
      structVersion: 1,
      id: request.id,
      hasToken: true,
      payload: confirmPayload,
    };
    const stepTwo = await this.sendCommandDirect(confirm);
    if (!stepTwo.ok) return failure(stepTwo.error);
    if (stepTwo.value.status === 'ok') return success(undefined);
    return failure(this.resultError(stepTwo.value));
  }

  private nextResult(commandId: DeviceCommandId): Promise<CommandResult> {
    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        cleanup();
        reject(new Error('timeout'));
      }, 3000);
      const listener = (result: CommandResult) => {
        if (result.commandId !== commandId) return;
        cleanup();
        resolve(result);
      };
      const cleanup = () => {
        clearTimeout(timeout);
        this.commandResultListeners.delete(listener);
      };
      this.commandResultListeners.add(listener);
    });
  }

  private waitForConfig(config: DeviceConfig, ms: number): Promise<void> {
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => reject(new Error('timeout')), ms);
      const listener = (value: DeviceConfig) => {
        if (configEquals(value, config)) {
          clearTimeout(timer);
          this.configListeners.delete(listener);
          resolve();
        }
      };
      this.configListeners.add(listener);
    });
  }

  private resultError(result: CommandResult): AppError {
    switch (result.status) {
      case 'range':
        return {
          kind: 'deviceRejected',
          message: `Устройство отклонило поле конфигурации (offset ${result.detail})`,
          fieldId: result.detail,
        };
      case 'notPaired':
        return AppErrors.notPaired;
      case 'storage':
        return { kind: 'storageError', message: 'Устройство не смогло сохранить настройки' };
      case 'tokenExpired':
        return { kind: 'tokenExpired', message: 'Время подтверждения истекло' };
      case 'hardware':
        return { kind: 'hardware', message: 'Устройство сообщило об аппаратной ошибке' };
      case 'notSupported':
        return { kind: 'unsupported', message: 'Функция отсутствует в этой версии прошивки' };
      default:
        return { kind: 'unknown', message: `Устройство отклонило операцию: ${result.status}` };
    }
  }
}

function delay(ms: number) {
  return new Promise((r) => setTimeout(r, ms));
}
